tachyon/zk/air/plonky3/base/multi_field32_conversions.h

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_ZK_AIR_PLONKY3_BASE_MULTI_FIELD32_CONVERSIONS_H_
+#define TACHYON_ZK_AIR_PLONKY3_BASE_MULTI_FIELD32_CONVERSIONS_H_
+
+#include <stdint.h>
+
+#include <array>
+#include <limits>
+
+#include "absl/types/span.h"
+
+#include "tachyon/base/containers/adapters.h"
+#include "tachyon/build/build_config.h"
+
+namespace tachyon::zk::air::plonky3 {
+
+template <typename BigF, typename SmallF>
+BigF Reduce(absl::Span<const SmallF> values) {
+  static_assert(SmallF::Config::kModulusBits <= 32);
+  static_assert(BigF::Config::kModulusBits > 64);
+
+  using BigInt = typename BigF::BigIntTy;
+  CHECK_LT(values.size(), BigInt::kLimbNums * 2);
+
+  BigInt ret;
+  for (size_t i = 0; i < values.size(); i += 2) {
+    uint32_t value = values[i].value();
+    if constexpr (SmallF::Config::kUseMontgomery) {
+      ret[i >> 1] = SmallF::Config::FromMontgomery(value);
+    } else {
+      ret[i >> 1] = value;
+    }
+    if (i < values.size() - 1) {
```

Comment:
Isn't this redundant since it is already checked by the for loop?

Reply:
I don't think so. This checks whether the size of the value is even or odd.